subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java

```diff
@@ -16,56 +16,219 @@
 
-    public AbstractMavenPublisher(LocalMavenRepositoryLocator mavenRepositoryLocator) {
-        this.mavenRepositoryLocator = mavenRepositoryLocator;
+abstract class AbstractMavenPublisher implements MavenPublisher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MavenPublisher.class);
+
+    private static final String POM_FILE_ENCODING = "UTF-8";
+    private final Factory<File> temporaryDirFactory;
+    private final XmlTransformer xmlTransformer = new XmlTransformer();
+
+    AbstractMavenPublisher(Factory<File> temporaryDirFactory) {
+        this.temporaryDirFactory = temporaryDirFactory;
     }
 
-    @Override
-    public void publish(MavenNormalizedPublication publication, MavenArtifactRepository artifactRepository) {
-        if (artifactRepository == null) {
-            LOGGER.info("Publishing to maven local repository");
-        } else {
-            LOGGER.info("Publishing to repository '{}' ({})", artifactRepository.getName(), artifactRepository.getUrl());
+    protected void publish(MavenNormalizedPublication publication, ExternalResourceRepository repository, URI rootUri, boolean localRepo) {
+        MavenProjectIdentity projectIdentity = publication.getProjectIdentity();
+        String groupId = projectIdentity.getGroupId().get();
+        String artifactId = projectIdentity.getArtifactId().get();
+        String version = projectIdentity.getVersion().get();
+
+        ModuleArtifactPublisher artifactPublisher = new ModuleArtifactPublisher(repository, localRepo, rootUri, groupId, artifactId, version);
+
+        if (isSnapshot(version)) {
+            ExternalResourceName snapshotMetadataPath = artifactPublisher.getSnapshotMetadataLocation();
+            Metadata snapshotMetadata = createSnapshotMetadata(publication, groupId, artifactId, version, repository, snapshotMetadataPath);
+
+            artifactPublisher.publish(snapshotMetadataPath, writeMetadataToTmpFile(snapshotMetadata, "snapshot-maven-metadata.xml"));
+
+            if (!localRepo) {
+                // Use the timestamped version for all published artifacts:
+                // The timestamped version is hidden deep in `Metadata.versioning.snapshotVersions`
+                artifactPublisher.artifactVersion = snapshotMetadata.getVersioning().getSnapshotVersions().get(0).getVersion();
+            }
+        }
+
+        if (publication.getMainArtifact() != null) {
+            artifactPublisher.publish(null, publication.getPackaging(), publication.getMainArtifact().getFile());
+        }
+        artifactPublisher.publish(null, "pom", publication.getPomArtifact().getFile());
+        for (MavenArtifact artifact : publication.getAdditionalArtifacts()) {
+            artifactPublisher.publish(artifact.getClassifier(), artifact.getExtension(), artifact.getFile());
         }
 
-        MavenPublishAction deployTask = createDeployTask(publication.getPackaging(), publication.getProjectIdentity(), mavenRepositoryLocator, artifactRepository);
```

Comment:
Now that we're no longer using a `MavenDeployAction` for remote deploys, we're also no longer using the `MavenDeployRetrier`.  That stuff seemed to be a little odd to me in the first place, so I'm not sure its a problem to remove it, but I want to make sure we're doing so consciously.  Thoughts?

Reply:
No integration tests failed when the functionality was removed, which to me means the PR should not have been merged. I think this functionality should be implemented at the `RepositoryTransport` level so that it works for Ivy & Maven and can be used for other features that do HTTP uploads.

I'll have a look, but I guess adding an integration test to master is probably the right place to start.

Reply:
So it turns out the `MavenDeployRetrier` wasn't really working, because we weren't wrapping network failures with the correct exception that would enable retry. If [fixed this on master](https://github.com/gradle/gradle/commit/d66492b1396ffa86506cd7a2e67616734da9d143) and added an integration test.

I'll now rebase my branch, expecting this new test to fail.